core/src/storage2/collections/stash/mod.rs

```diff
@@ -0,0 +1,549 @@
+            return None
+        }
+        // Precompute prev and next vacant entires as we might need them later.
+        // Due to borrow checker constraints we cannot have this at a later stage.
+        let (prev, next) = if let Some(index) = self.last_vacant_index() {
+            let root_vacant = self
+                .entries
+                .get(index)
+                .expect("last_vacant must point to an existing entry")
+                .try_to_vacant()
+                .expect("last_vacant must point to a vacant entry");
+            // Form the linked vacant entries in a way that makes it more likely
+            // for them to refill the stash from low indices.
+            if at < index {
+                // Insert before root if new vacant index is smaller than root.
+                (root_vacant.prev, index)
+            } else if at < root_vacant.next {
+                // Insert between root and its next vacant entry if smaller than
+                // current root's next index.
+                (index, root_vacant.next)
+            } else {
+                // Insert before root entry if index is greater. But we won't
+                // update the new element to be the new root index in this case.
+                (root_vacant.prev, index)
+            }
+        } else {
+            // Default prev and next to the given at index.
+            // So the resulting vacant index is pointing to itself.
+            (at, at)
+        };
+        let entry_mut = self.entries.get_mut(at).expect("index is out of bounds");
+        if entry_mut.is_vacant() {
+            // Early return if the taken entry is already vacant.
+            return None
+        }
+        // At this point we know that the entry is occupied with a value.
+        let new_vacant_entry = Entry::Vacant(VacantEntry { prev, next });
+        let taken_entry = core::mem::replace(entry_mut, new_vacant_entry);
+        // Update links from and to neighbouring vacant entries.
+        if prev == next {
+            // Previous and next are the same so we can update the vacant
+            // neighbour with a single look-up.
+            if let Some(entry) = self
+                .entries
+                .get_mut(next)
+                .map(Entry::try_to_vacant_mut)
+                .expect("`next` must point to an existing entry at this point")
+            {
+                entry.prev = at;
+                entry.next = at;
+            }
+        } else {
+            // Previous and next vacant entries are different and thus need
+            // different look-ups to update them.
+            if let Some(entry) = self
+                .entries
+                .get_mut(prev)
+                .map(Entry::try_to_vacant_mut)
+                .expect("`prev` must point to an existing entry at this point")
+            {
+                entry.next = at;
+            }
+            if let Some(entry) = self
```

Comment:
It looks like `next`and `prev` should always be be vacant, because they are one of: `root_vacant.prev`, `root_vacant.next`, `index` (from `last_vacant_index`), or `at` which we have just assigned as vacant. 

If I'm right we could replace the `if let Some(entry)` here and above with another `expect`.

Reply:
yep, done!